AzureMonitor: Wait for credentials to be ready

aangelisc opened the pull request:
This PR exposes the `config` property allowing Cypress configuration to be specified as a part of the scenario.

Also updates the Azure Monitor E2E tests to retry the datasource configuration on failure.

This PR may need to be backported to `v9.3.x` to reduce test flakiness.

aangelisc committed:
Add ability to retry e2e test scenarios

andresmgot commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -45,6 +46,7 @@ function provisionAzureMonitorDatasources(datasources: AzureMonitorProvision[])
       e2eSelectors.configEditor.defaultSubscription.input().find('input').type('datasources{enter}');
     },
     expectedAlertMessage: 'Successfully connected to all Azure Monitor endpoints',
+    timeout: 10000,
   });
```
another possibility is adding here a try-catch if we know this is the specific step that may fail. That way we click the "Save" button twice rather than creating a new data source.

aangelisc replied:
Try-catch statements don't work in Cypress :/ Cypress is designed to prevent you from retrying specific steps (that was my initial attempt, see details [here](https://docs.cypress.io/guides/core-concepts/conditional-testing#Error-Recovery)). I agree with all of your comments above, particularly about hiding legitimate errors. I'll have more of a think on this and maybe implement some uncaught exception handling.

andresmgot reviewed (commented):
Thanks @aangelisc! In general, I think it's risky to retry whole tests (see my comment). Let me know what you think

aangelisc committed:
Add wait and remove retries

aangelisc commented:
I've updated this with a `wait` to attempt to give the credentials time to be ready. I've also removed the `retries` however I think it's still beneficial to expose the `config` parameter so I'm leaving that in for now.

aangelisc committed:
Revert retry logic and elaborate on 15s wait time

andresmgot commented on e2e/cloud-plugins-suite/azure-monitor.spec.ts:
```diff
@@ -20,10 +20,11 @@ type AzureMonitorConfig = {
 type AzureMonitorProvision = { datasources: AzureMonitorConfig[] };
 
-const dataSourceName = `Azure Monitor E2E Tests - ${uuidv4()}`;
+let dataSourceName = '';
```
nit: you don't need this change now either :P

aangelisc replied:
Oops, thanks!

andresmgot reviewed (approved):
LGTM! Thanks a lot for your efforts making these tests stable!

aangelisc committed:
No need to update datasource name now
